Ensure all async writes are completed before realm is disposed

peppy opened the pull request:
Fixes following failure:

```csharp
TearDown : System.AggregateException : One or more errors occurred. (Cannot access a disposed object.
Object name: 'RealmAccess'.)
  ----> System.ObjectDisposedException : Cannot access a disposed object.
Object name: 'RealmAccess'.
--TearDown
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at osu.Framework.Extensions.TaskExtensions.WaitSafely(Task task)
   at osu.Framework.Testing.TestScene.checkForErrors()
   at osu.Framework.Testing.TestScene.RunTestsFromNUnit()
--ObjectDisposedException
   at osu.Game.Database.RealmAccess.getRealmInstance() in /opt/buildagent/work/ecd860037212ac52/osu.Game/Database/RealmAccess.cs:line 562
   at osu.Game.Database.RealmAccess.<>c__DisplayClass33_0.<<WriteAsync>b__0>d.MoveNext() in /opt/buildagent/work/ecd860037212ac52/osu.Game/Database/RealmAccess.cs:line 400
--- End of stack trace from previous location ---
   at osu.Framework.Platform.GameHost.<>c__DisplayClass128_0.<abortExecutionFromException>b__0()
   at osu.Framework.Threading.ScheduledDelegate.RunTaskInternal()
   at osu.Framework.Threading.Scheduler.Update()
   at osu.Framework.Threading.GameThread.processFrame()
   at osu.Framework.Threading.GameThread.RunSingleFrame()
   at osu.Framework.Platform.ThreadRunner.RunMainLoop()
   at osu.Framework.Platform.GameHost.windowUpdate()
   at osu.Framework.Platform.GameHost.Run(Game game)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state)
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
```

peppy committed:
Remove unnecessary `Task.Run` workaround in tests

peppy committed:
Ensure all async writes are completed before realm is disposed

peppy committed:
Throw immediately if attempting to `WriteAsync` after disposed

peppy committed:
Add test coverage of realm async writes

bdach commented on osu.Game/Database/RealmAccess.cs:
```diff
@@ -388,26 +388,46 @@ public void Write(Action<Realm> action)
             {
-                total_writes_async.Value++;
+                // CountdownEvent will fail if already at zero.
+                if (!pendingAsyncWrites.TryAddCount())
+                    pendingAsyncWrites.Reset(1);
```
`.TryAddCount(int)` is [documented as not actually being thread-safe](https://docs.microsoft.com/en-us/dotnet/api/system.threading.countdownevent.tryaddcount?view=net-6.0#system-threading-countdownevent-tryaddcount(system-int32)), which is a problem given that the `.Dispose()` usage site is using it without further locking. Any reason this isn't just [a simple `.AddCount()`?](https://docs.microsoft.com/en-us/dotnet/api/system.threading.countdownevent.addcount?view=net-6.0#system-threading-countdownevent-addcount) That method doesn't check if the count is already zero, at least not as far as I can tell from xmldoc.

peppy replied:
Literally no reason. I just see a `Try` method and assume it is the better of the two, based on prior knowledge. Feel free to change it if it works and passes tests, seems good.

peppy replied:
`AddCount` does seem to fail:

![JetBrains Rider 2022-06-28 at 08 11 20](https://user-images.githubusercontent.com/191335/176128463-bd13c6eb-e6d5-40e0-8ed2-535122575aa3.png)

bdach commented on osu.Game/Database/RealmAccess.cs:
```diff
@@ -910,6 +930,9 @@ void restoreOperation()
 
         public void Dispose()
         {
+            if (!pendingAsyncWrites.Wait(10000))
+                Logger.Log("Realm took too long waiting on pending async writes", level: LogLevel.Error);
```
This lock ordering is giving me the creeps. `WriteAsync()` takes `realmLock` first and then adds count to `pendingAsyncWrites`, while this waits on the latter first and only ever then takes the former. Not to mention that `pendingAsyncWrites` is being signaled from TPL. The timeout ensures that this won't deadlock, but I don't know that this will never timeout, nor that it will work properly.

At least in my head, if this `.Wait()` is done after taking `realmLock`, then there's the guarantee that no new writers can add a count to the `CountdownEvent` (because they'd have to take `realmLock` first).

bdach replied:
Not sure I see it. If I apply

```diff
diff --git a/osu.Game/Database/RealmAccess.cs b/osu.Game/Database/RealmAccess.cs
index 7a9561fc5e..e3edec254b 100644
--- a/osu.Game/Database/RealmAccess.cs
+++ b/osu.Game/Database/RealmAccess.cs
@@ -930,11 +930,11 @@ void restoreOperation()
 
         public void Dispose()
         {
-            if (!pendingAsyncWrites.Wait(10000))
-                Logger.Log("Realm took too long waiting on pending async writes", level: LogLevel.Error);
-
             lock (realmLock)
             {
+                if (!pendingAsyncWrites.Wait(10000))
+                    Logger.Log("Realm took too long waiting on pending async writes", level: LogLevel.Error);
+
                 updateRealm?.Dispose();
             }
```

all test cases from files touched in this diff still pass.

peppy replied:
Probably just good luck. Have a look at `WriteAsync` -> `getRealmInstance` -> `realmLock`

I’ll try and add better test coverage of this tomorrow…

peppy replied:
I know this is going to likely increase the complexity of this PR, but I've removed `realmLock` completely. I can't understand why it would be required. I've also run tests through quite a few iterations with no failures.

Let me know your thoughts. Probably worth taking a good look through the class and making sure I'm not missing something. If we need more locking than the `realmRetrievalLock`, it needs to be named and documented far better.

bdach replied:
Right now, as I see it, the primary things that `realmLock` is protecting are probably the `updateRealm` singleton and the `customSubscriptionResetMap`, which is a non-concurrent collection. Most private places are guaranteed to be taking that lock on the update thread (`ensureUpdateRealm()`, `RegisterCustomSubscription+unsubscribe()`, which is sent/posted, `registerSubscription()` which is either called from `ensureUpdateRealm()`, or sent/posted).

The ones that worry me more are the public ones: `RegisterForNotifications<T>()`, `BlockAllOperations()`, `Dispose()`. They *look* like there should be no thread contention on them (always should be getting called from update), but I'm not 100% sure if I'm honest. Maybe fine to stick `ThreadSafety.IsUpdateThread` checks everywhere and call it a day?

peppy replied:
Yeah, I think enforcing update thread is much better than locking.. until we need otherwise.

peppy commented:
The `realmLock` is confusing as hell, to the point I'm not sure it's even correct. I think the goal was to guard against usage after disposal (ie. it should be called `realmDisposalLock`) but I'm still confirming this.

Hold off reviewing this until I figure that out.

peppy committed:
Remove probably redundant `realmLock`

As far as I can tell all accesses are safe due to update thread
guarantees. The only weird one may be async writes during a
`BlockAllOperations`, but the `Compact` loop should handle this quite
amicably.

peppy committed:
Add test coverage of async writes during a blocking operation

peppy committed:
Add back thread safety and locking as required

peppy committed:
Merge branch 'master' into realm-fix-async-write-after-disposal

peppy committed:
Add one more lock to appease CI

peppy committed:
Merge branch 'master' into realm-fix-async-write-after-disposal

peppy commented:
This should be ready for re-review now, in case that wasn't clear.

smoogipoo reviewed (approved):
I've gone through this and it looks good to me.
